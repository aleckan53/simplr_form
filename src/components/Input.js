import React from 'react';
import useInput from 'hooks/useInput';

const Input = ({ label, id, type = 'text', children, hint, optional }) => {
  const { valid, onChange, invalidMsg } = useInput(id, type, optional);

  // conditionally renders invalid msg
  const showMsg = !valid && type !== 'checkbox';

  return <>
    <label htmlFor={id}>{label} {hint && <span>{hint}</span>}</label>
    <input id={id} className={`input ${type}`} type={type} onChange={onChange} data-valid={valid}/>
    {showMsg && <div className='invalidMsg'><p>{invalidMsg}</p></div>}
    {children}
  </>
};

export default Input;