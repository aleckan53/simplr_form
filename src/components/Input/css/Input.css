.input {
  -webkit-appearance: none;
     -moz-appearance: none;
          appearance: none;
  border-radius: 0;
  width: 100%;
  height: 2rem;
  padding: .2rem .5rem;
  font-size: .9rem;
  color: black;
  border-style: solid;
  border-width: 1px;
  border-color: var(--inputBorder);
  background-color: var(--inputBg);
}

.input:focus {
  outline-color: var(--mainColor);
  outline-style: solid;
  outline-width: 2px;
  background: #fff;
}

.input[data-valid='false']:not(.checkbox) {
  border: 1px solid red;
}

.input.checkbox {
  width: auto;
  display: inline-block;
  -webkit-appearance: none;
     -moz-appearance: none;
          appearance: none;
  width: 1rem;
  height: 1rem;
  border: 1px solid var(--labelColor);
  cursor: pointer;
}

.input.checkbox:checked {
  background-image: url('data:image/svg+xml,%3Csvg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="0.181in" height="0.139in"%3E%3Cpath fill-rule="evenodd" fill="rgb(255, 255, 255)" d="M13.000,1.249 L11.419,-0.002 L5.180,7.255 L1.284,4.257 L0.001,5.790 L5.475,10.002 L13.000,1.249 Z"/%3E%3C/svg%3E');
  background-color: #4b4b4b;
  background-position: 2px 2px, 0 0;
  background-repeat: no-repeat;
  background-size: contain;
}

.invalidMsg {
  position: relative;
}

.invalidMsg p {
  position: absolute;
  left: 0;
  color: red;
  font-size: 12px;
}

label {
  display: block;
  color: var(--labelColor);
  margin-bottom: .2rem;
  font-size: .9rem;
}

label:not(:first-of-type) {
  margin-top: 1rem;
}

label span {
  font-size: .6rem;
  color: var(--lightGray);
}

@media only screen and (max-width: 374px) {
  label span {
    font-size: 8px;
  }
}

label[data-for='checkbox'] {
  font-weight: 700;
  font-size: .8rem;
}

label[data-for='checkbox']:last-of-type {
  margin-top: .1rem;
}

input:-webkit-autofill,
input:-webkit-autofill:hover,
input:-webkit-autofill:focus,
input:-webkit-autofill:active {
  -webkit-box-shadow: 0 0 0 30px var(--inputBg) inset !important;
          box-shadow: 0 0 0 30px var(--inputBg) inset !important;
}
/*# sourceMappingURL=Input.css.map */