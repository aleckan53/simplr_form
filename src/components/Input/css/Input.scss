@mixin font($size: 1rem, $color: yellow, $weight: 400) {
  color: $color;
  font-size: $size;
  font-weight: $weight
}

.input {
  @include font(.9rem, black);
  appearance: none;
  border-radius: 0;
  border: 1px solid var(--inputBorder);
  width: 100%;
  height: 2rem;
  padding: .2rem .5rem;
  background-color: var(--inputBg);

  &:focus {
    outline-color: var(--mainColor);
    outline-style: solid;
    outline-width: 2px;
    background: #fff;
  }

  &[data-valid='false']:not(.checkbox) {
    border: 1px solid red;
  }

  &.checkbox {
    width: auto;
    display: inline-block;
    appearance: none;
    width: 1rem;
    height: 1rem;
    border: 1px solid var(--labelColor);
    cursor: pointer;

    &:checked {
      background-image: url('data:image/svg+xml,%3Csvg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="0.181in" height="0.139in"%3E%3Cpath fill-rule="evenodd" fill="rgb(255, 255, 255)" d="M13.000,1.249 L11.419,-0.002 L5.180,7.255 L1.284,4.257 L0.001,5.790 L5.475,10.002 L13.000,1.249 Z"/%3E%3C/svg%3E');
      background-color: #4b4b4b;
      background-position: 2px 2px, 0 0;
      background-repeat: no-repeat;
      background-size: contain;
    }
  }
}

.invalidMsg {
  position: relative;

  p {
    position: absolute;
    left: 0;
    @include font(.7rem, red);
  }
}

label {
  display: block;
  color: var(--labelColor);
  margin-bottom: .2rem;
  font-size: .9rem;
  
  &:not(:first-of-type) {
    margin-top: 1rem;
  }

  span {
    @include font(.6rem, var(--lightGray));
    @media only screen and (max-width: 374px) {
      font-size: 8px;
    }
  }

  &[data-for='checkbox'] {
    font-weight: 700;
    font-size: .8rem;
    @include font(.8rem, #000, 700);

    &:last-of-type {
      margin-top: .7rem;
    }
  }
}

input:-webkit-autofill,
input:-webkit-autofill:hover, 
input:-webkit-autofill:focus, 
input:-webkit-autofill:active  {
  box-shadow: 0 0 0 30px var(--inputBg) inset !important;
}