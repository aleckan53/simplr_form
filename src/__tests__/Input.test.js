import React from 'react';
import { mount } from 'enzyme';
import Input from '../components/Input';
import toJson from 'enzyme-to-json';

describe('Input', () => {
  let wrapper;

  afterEach(() => {
    wrapper.unmount();
  });

  it('matches snapshot', () => {
    wrapper = mount(<Input/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders props correctly', () => {
    wrapper = mount(
      <Input 
        label='Test' 
        id='testId' 
        type='text' 
        hint='testHint'/>
    );

    const input = wrapper.find('input');
    const label = wrapper.find('label');
    const hint = label.find('span');
    const invalidMsg = wrapper.find('.invalidMsg').find('p');

    expect(hint.text()).toBe('testHint');
    expect(label.text()).toBe('Test testHint');
    expect(input.prop('type')).toBe('text');
    expect(label.prop('htmlFor')).toBe('testId');
    expect(input.prop('id')).toBe('testId');
    expect(invalidMsg.text()).toBeFalsy();
  });

  it('doesn\'t render validation msg if type="checkbox"', () => {
    wrapper = mount(<Input type='checkbox'/>);
    expect(wrapper.children()).not.toContain('.invalidMsg');
  });

  it('renders children if given', () => {
    wrapper = mount(<Input><h1>Test</h1></Input>);
    const testDiv = wrapper.find('h1');

    expect(testDiv.text()).toBe('Test');
  });
});