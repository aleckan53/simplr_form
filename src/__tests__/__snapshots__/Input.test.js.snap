// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches snapshot 1`] = `
<Input>
  <label
    data-for="text"
  >
     
  </label>
  <input
    className="input text"
    data-valid={null}
    onChange={[Function]}
    type="text"
  />
  <div
    className="invalidMsg"
  >
    <p />
  </div>
</Input>
`;
